flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/CombineRecordReader.java

```diff
@@ -0,0 +1,180 @@
+ * the same time to determine if previous one can be returned. So the reader only accepts wrapped
+ * reader with {@link RecordIterator#supportPrevious()}.
+ */
+public abstract class CombineRecordReader implements RecordReader<KeyValue> {
+
+    private final RecordReader<KeyValue> reader;
+    private final Comparator<RowData> userKeyComparator;
+    private final RowDataSerializer userKeySerializer;
+    private final RowDataSerializer valueSerializer;
+
+    private KeyValue candidate;
+
+    public CombineRecordReader(
+            RecordReader<KeyValue> reader,
+            Comparator<RowData> userKeyComparator,
+            RowDataSerializer userKeySerializer,
+            RowDataSerializer valueSerializer) {
+        this.reader = reader;
+        this.userKeyComparator = userKeyComparator;
+        this.userKeySerializer = userKeySerializer;
+        this.valueSerializer = valueSerializer;
+    }
+
+    @Nullable
+    @Override
+    public RecordIterator<KeyValue> readBatch() throws IOException {
+        RecordIterator<KeyValue> iterator = reader.readBatch();
+        if (iterator != null) {
+            return new CombineRecordIterator(iterator);
+        } else {
+            if (candidate != null) {
+                iterator = new SingletonRecordIterator<>(candidate);
+                candidate = null;
+                return iterator;
+            }
+        }
+        return null;
+    }
+
+    @Override
+    public void close() throws IOException {
+        this.reader.close();
+    }
+
+    public abstract KeyValue combine(KeyValue from, KeyValue to);
+
+    private class CombineRecordIterator implements RecordIterator<KeyValue> {
+
+        private final RecordIterator<KeyValue> iterator;
+
+        private KeyValue current;
+
+        private CombineRecordIterator(RecordIterator<KeyValue> iterator) {
+            checkArgument(iterator.supportPrevious());
+            this.iterator = iterator;
+        }
+
+        @Override
+        public boolean advanceNext() throws IOException {
+            if (candidate == null) {
+                boolean success = iterator.advanceNext();
+                if (!success) {
+                    return false;
+                }
+                candidate = iterator.current();
+            }
+
+            while (true) {
+                if (iterator.advanceNext()) {
+                    // change key, we can output candidate
+                    if (userKeyComparator.compare(iterator.current().key(), candidate.key()) != 0) {
+                        current = candidate;
+                        candidate = iterator.current();
+                        return true;
+                    } else {
+                        candidate = combine(candidate, iterator.current());
+                    }
+                } else {
+                    candidate = candidate.copy(userKeySerializer, valueSerializer);
+                    return false;
```

Comment:
Why return `false`? If I'm not mistaken this branch is to produce the last record and should return `true`. Also this should be `current = candidate` and there is no need for `copy`?

Reply:
OK I understand. Because the input is split into batches and there might be more records with the same key in the next batch so we can't conclude here.

Please add more comments to explain the logic. This is somewhat confusing.